src/PieChart/PieArcSeries/PieArcLabel.tsx

```diff
@@ -24,7 +32,8 @@ export class PieArcLabel extends PureComponent<PieArcLabelProps> {
     fontFill: '#8F979F',
     fontSize: 11,
     fontFamily: 'sans-serif',
-    padding: '.35em'
+    padding: '.35em',
+    height: 11
```

Comment:
Is height not calculated automatically - I would prefer not to pass this as its prone to errors.

Reply:
Previously this value was hard coded in `calculateLabelPositions` as `const minDistance = 15;` (11 line height + 4px margin). This value can not be used to properly align custom labels, so I've made it possible to change defaults.

I think that it is possible to calculate height automatically in case when user uses custom labels, but it will make things really hard. The `calculateLabelPositions` uses item height to ensure that the labels won't overlap. If this height will be random for each item, it will be really hard. Especially because currently label positions are calculated before the labels are rendered.

I think it's better to require users to specify the dimensions they need (if you'll use icons, you'll be know the label size in advance). This will result in more predicted chart appearance. It's better to go with explicit label sizes till we'll have feature request with arguments why dynamic sizing is really needed.

BTW. The new way of providing label size can also help with the default label styles, because now you can provide `width` for label to ensure, that the chart will reserve this amount of place for labels. Previously you had to change horizontal margins in order to reserve some place for labels.

Reply:
I have similar code in a few other spots:

- https://github.com/reaviz/reaviz/blob/master/src/VennDiagram/VennOuterLabel.tsx#L48
- https://github.com/reaviz/reaviz/blob/master/src/ScatterPlot/ScatterSeries/ScatterPoint.tsx#L138

and demos:

- https://reaviz.io/?path=/story/charts-venn-diagram--label-icons
- https://reaviz.io/?path=/story/charts-scatter-plot-linear--symbols
- https://reaviz.io/?path=/story/charts-scatter-plot-radial--symbols
- https://reaviz.io/?path=/story/charts-line-chart-circle-series--shapes

Reply:
How about we refactor this to be more generic so you can pass anything such as a image, svg, component and allow users to specify the width and height.

For labels I think its extremely error prone for users to have to pass widths for a variety of reasons:

1. No one knows what the width of text actually is - a lot of people use char count but thats not valid. I actually have a measure system internally to measure this.
2. Its extremely common for users to not know what the text will be until the server returns it, making it difficult as well to measure.

I think for icons/elements its OK though since those are typically know things.

Reply:
Current implementation should allow you to pass any contents. In storybook demo I'm using html with flexbox layout and svg inside. If the user want the text only label, he can use the old way with providing the strings. But even in this case the user may want to explicitly specify maximum allowed label width.

I want to share the points, why I have decided to go with explicit width/height. In the first version of this feature, where I've just patched `reaviz` to be able to use icons without waiting for PR to accept (it was in the old reaviz repo) and here the issues, that was with that first implementation:

Because I've decided to allow dynamic label sizes, I've used `foreignObject` with `width` and `height` equal `0` and `overflow: visible`. As a result I've also added a div with `position: absolute` inside to compensate the effect of zero size. 

After that it started to be possible to use any label content of any size. But with some new issues: because of position absolute, the content can not properly stretch anymore, so in order to get a text inline, you should provide `white-space: nowrpap` or it will print each word from the new line. I think this is a bad experience and it will be more difficult for the end user, than explicit sizes.

The next issue:

![image](https://user-images.githubusercontent.com/2536916/97774467-ac82dd80-1b60-11eb-8711-db5d79d4bf7b.png)

If the size has no bounds, it's probably impossible to ensure, that my labels will fit the screen. To fix example above I'll should end with some fixed horizontal margin for chart (which should include actual chart margins and width reserved for labels) and additionally I should add fixed width for all of my labels to stop them endlessly stretching. This fix can not be applied programmatically by reaviz because even if we measure label size we don't know the threshold to decide which label is too long. If we ask user to decide how much space to allocate for the labels we will produce much nicer output than when we will try to calculate (guess) some threshold.

When the sizes are explicit I can decide how much space I want to allocate for each label (this does not mean that contents must have the exact size) and then style it's contents so that they will look more pleasant, e.g.:

![image](https://user-images.githubusercontent.com/2536916/97774581-b8bb6a80-1b61-11eb-9d09-23a974e6979f.png)

In this case I can add "…" and crop the long text instead of showing the icons offscreen.

Reply:
> How about we refactor this to be more generic so you can pass anything such as a image, svg, component and allow users to specify the width and height.

Do you mean to refactor all other charts to allow any (html, img etc) content for labels? I've thought to do so, but I think it should be done one by one as separate PRs

Reply:
I would really rather NOT pass a height. Really the height only comes into play when there is a foreign object so maybe let that specification happen there and then we just measure it when it comes back?

For the label overflow situation, I did something similar for the Venn diagram that you might want to checkout. https://github.com/reaviz/reaviz/blob/master/src/VennDiagram/VennLabel.tsx#L84

Typically what I do when I hit a problem like this is look at how other frameworks do it - take a look at some of the other frameworks I list in the docs under "Why".

Reply:
> take a look at some of the other frameworks I list in the docs under "Why".

Almost all libs have very poor label customization with very poor positioning. So we can not use their experience. Victory has some height-related props (lineHeight, capHeight). I've used `vx`, but it delegates almost everything to developer. Therefore there is no controls for height. [Here](https://williaster.github.io/data-ui/?selectedKind=xy-chart&selectedStory=TickLabelPlayground&full=0&addons=0&stories=1&panelRight=0) is really smart implementation of labels and it requires you to set width. The user must provide how much place to reserve for labels. We can not decide about this width based on label content, because some labels may take the whole chart available width.

Reply:
Current implementation does not require user to provide width/height. It is opt-in. If you use simple textual labels, you probably won't need to specify width/height, because default positioning algorithm reserves 15 pixels of height and 1/3 of available radius for each label (this is exactly the same logic as it was before my PR).

If the user will need to use foreignObject, he will (most likely) need to specify different height (or even width), because the default one won't fit

So in most cases the user will provide height/width only when he uses the foreignObject and in any case we do not require the user to provide those sizes, they are completely optional

Reply:
In future we may start using `wrapText` as in [Venn diagram](https://github.com/reaviz/reaviz/blob/master/src/VennDiagram/VennLabel.tsx#L84), but take a not, that venn diagram actually has some sort of pre-specified sizes — `const width = data?.circles?.[0]?.radius;` In case with PieChart the user should be able to specify how much width he wants to reserve for labels, because the labels are drawn not inside of the circles, but outside.

When I will implement support for icons inside PieArc I won't ask user for width/height, because that icons will be placed inside arc, where I exactly know how much free space do I have.

Reply:
Got it - ya i think for now this is fine and we can iterate on it more.